File: 12-triangular-number/main.c
#include "math.h"
#include "stdio.h"
#include "time.h"

long long firstNum(int moreThan) {
    long long number = 0;
    long long delta = 1;
    while (1) {
        number = number + delta;
        double num_sqrt = sqrt(number);
        delta++;
        int count = 0;
        for (long long i = 1; i < num_sqrt; i++) {
            if (number % i == 0) {
                count++;
            }
        }
        count = 2*count;
        if (number == num_sqrt / 1) {
            count++;
        }
        if (count > moreThan) {
            return number;
        }
    }
}

int main() {
    clock_t start = clock();
    long long result = firstNum(500);
    clock_t end = clock();
    printf("time taken: %f\n", ((double) (end - start)) / CLOCKS_PER_SEC);
}

File: 12-triangular-number/main.py
import math
from util import timing

@timing
def first_num(more_than: int) -> int:
    number = 0
    delta = 1
    while True:
        number += delta
        delta += 1
        count = 0
        num_sqrt = math.sqrt(number)
        for i in range(1, math.ceil(num_sqrt)):
            if number % i == 0:
                count += 1
        count *= 2
        if num_sqrt % 1 == 0:
            count += 1 
        if count > more_than:
            return number

print(f"The first triangle number with over 500 divisors is:\n{first_num(500)}")
